src/progress.rs

```diff
@@ -167,60 +290,109 @@ impl ProgressSet {
         self.progress.iter_mut()
     }
 
-    /// Adds a voter node
-    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        // If the progress exists already this is in error.
-        if self.progress.contains_key(&id) {
-            // Determine the correct error to return.
-            if self.learner_ids().contains(&id) {
-                return Err(Error::Exists(id, "learners"));
-            }
+    /// Adds a voter to the group.
+    ///
+    /// # Errors
+    ///
+    /// * `id` is in the voter set.
+    /// * `id` is in the learner set.
+    /// * There is a pending membership change.
+    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<()> {
+        debug!("Inserting voter with id {}.", id);
+
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
             return Err(Error::Exists(id, "voters"));
+        } else if self.is_in_membership_change() {
+            return Err(Error::ViolatesContract(
+                "There is a pending membership change.".into(),
+            ));
         }
+
         self.configuration.voters.insert(id);
         self.progress.insert(id, pr);
+
         self.assert_progress_and_configuration_consistent();
         Ok(())
     }
 
-    /// Adds a learner to the cluster
-    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        // If the progress exists already this is in error.
-        if self.progress.contains_key(&id) {
-            // Determine the correct error to return.
-            if self.learner_ids().contains(&id) {
-                return Err(Error::Exists(id, "learners"));
-            }
+    /// Adds a learner to the group.
+    ///
+    /// # Errors
+    ///
+    /// * `id` is in the voter set.
+    /// * `id` is in the learner set.
+    /// * There is a pending membership change.
+    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<()> {
+        debug!("Inserting learner with id {}.", id);
+
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
             return Err(Error::Exists(id, "voters"));
+        } else if self.is_in_membership_change() {
+            return Err(Error::ViolatesContract(
+                "There is a pending membership change.".into(),
+            ));
         }
+
         self.configuration.learners.insert(id);
         self.progress.insert(id, pr);
+
         self.assert_progress_and_configuration_consistent();
         Ok(())
     }
 
     /// Removes the peer from the set of voters or learners.
-    pub fn remove(&mut self, id: u64) -> Option<Progress> {
-        self.configuration.voters.remove(&id);
+    ///
+    /// # Errors
+    ///
+    /// * There is a pending membership change.
+    pub fn remove(&mut self, id: u64) -> Result<Option<Progress>> {
```

Comment:
Looks like this function should be named `remove_voter`. Compare with `insert_voter`, `insert/remove_learner`.

Reply:
Wait, this function does both learners and voters. Hm.

Reply:
This removes voters or learners, so we've need to provide `remove_learner` as well. :( So keeping it just `remove` is simpler.

Reply:
The docs should probably indicate this works for both voters and learners, since it's asymmetrical with the insert methods.

Reply:
Oh, wait again, the docs say that.

Reply:
Yup!

```
/// Removes the peer from the set of voters or learners.
```